Домашнее задание №7

MAMKUH6OPIIIUK opened the pull request:
Выполнил ДЗ №7.
Отличия от реализации из ДЗ№6:
1. Репозитории:
- убраны все реализации, использующие `EntityManager`
- интерфейсы теперь просто наследуются от `JpaRepository`. Для сохранения решения N+1 перенёс некоторые jpql запросы из старых реализаций в аннотации `@Query` над методами интерфейсов
- объявления большей части методов убраны, т.к. есть в родительских интерфейсах. Оставил только кастомные методы и методы, возвращающие списки (для указания конкретного типа возвращаемой коллекции)
- метод` findByIds` у `GenreRepository` переименован в `findByIdIn`
2. Тесты репозиториев:
- Убрана аннотация `@Import` (в которой импортировалась конкретная тестируемая реализация)
- репозитории теперь инжектятся по интерфейсу.

Всё остальное завелось без изменений 😄

MAMKUH6OPIIIUK committed:
ДЗ №7. Initial commit

MAMKUH6OPIIIUK committed:
Выполнил ДЗ №7

MAMKUH6OPIIIUK committed:
Добавил аннотации `@Repository` над интерфейсами репозиториев (хоть это, видимо, и не обязательно)

MAMKUH6OPIIIUK committed:
Убрал переопределения findAll там, где они не нужны.
И всё-таки убрал аннотации Repository

vlad-iv commented on spring-hw-07/pom.xml:
```diff
@@ -0,0 +1,125 @@
+        <checkstyle.config.url>https://raw.githubusercontent.com/OtusTeam/Spring/master/checkstyle.xml</checkstyle.config.url>
+        <minimal.maven.version>3.3.9</minimal.maven.version>
+        <spring.shell.version>3.0.0</spring.shell.version>
+        <h2.version>2.1.214</h2.version>
```
Предлагаю версию подтягивать из spring-boot-starter-parent (можно удалить)

MAMKUH6OPIIIUK replied:
Убрал свойство для версии h2

vlad-iv commented on spring-hw-07/pom.xml:
```diff
@@ -0,0 +1,125 @@
+        <dependency>
+            <groupId>com.h2database</groupId>
+            <artifactId>h2</artifactId>
+            <version>${h2.version}</version>
```
Можно удалить

MAMKUH6OPIIIUK replied:
Спасибо! Удалил версию в dependency - оказывается, есть в dependencyManagement у spring-boot-dependencies - родителя всех стартеров. Не знал..

vlad-iv commented on spring-hw-07/src/main/java/ru/otus/spring/homework/oke/model/Book.java:
```diff
@@ -0,0 +1,58 @@
+@Setter
+@Entity
+@Table(name = "books")
+public class Book {
```
Предлагаю добавить @NamedEntityGraph для загрузки автора

MAMKUH6OPIIIUK replied:
Добавил, тоже удобная штука 👍

vlad-iv commented on spring-hw-07/src/main/java/ru/otus/spring/homework/oke/repository/BookRepository.java:
```diff
@@ -0,0 +1,17 @@
+    List<Book> findByAuthorId(Long authorId);
+
+    @Override
+    @Query("select distinct b from Book b join fetch b.author")
```
Предлагаю заменить на @EntityGraph
так же может иметь смысл добавить для метода findById

MAMKUH6OPIIIUK replied:
Заменил. Думал, что Query нагляднее, т.к. сразу в репозитории видно, как подтягиваются данные. Но у графов тоже много плюсов 👍 
С findById - согласен, сделал его переопределение с entitygraph тоже

vlad-iv commented on spring-hw-07/src/main/java/ru/otus/spring/homework/oke/service/BookService.java:
```diff
@@ -0,0 +1,18 @@
+public interface BookService {
+    BookResponseDto create(BookRequestDto bookRequestDto);
+
+    void update(Long id, BookRequestDto bookRequestDto);
```
Можно добавить поле Long id в BookRequestDto

MAMKUH6OPIIIUK replied:
Добавил.. 
Не добавлял затем, чтобы не выставлять потом в потенциальном rest api поле, которое клиенту не надо заполнять совсем при create, и надо заполнять как `@PathVariable`, а не часть тела запроса, при update. Хотя, возможно, эти мои предположения далеки от best practices реализации crud через rest,  буду ждать грядущих занятий 😄

vlad-iv commented on spring-hw-07/src/main/java/ru/otus/spring/homework/oke/service/BookServiceImpl.java:
```diff
@@ -0,0 +1,135 @@
+    @Override
+    @Transactional
+    public void update(Long id, BookRequestDto bookRequestDto) {
+        this.validateBook(id);
```
Предлагаю следующий подход
1. по id получаем книгу
2. получаем автора и жанры (уже есть)
3. обновляем данные у полученной книги
4. сохраняем (лучше оставить, но вызов не обязательный)

MAMKUH6OPIIIUK replied:
Сделал, но не без сложностей.. 
Если просто у полученной книги вызвать setGenres с новой коллекцией жанров, то получается вот такое в логах:
`Hibernate: update books set author_id=?, description=?, title=? where id=?`
`Hibernate: delete from books_genres where book_id=?`
`Hibernate: insert into books_genres (book_id, genre_id) values (?, ?)`
`Hibernate: insert into books_genres (book_id, genre_id) values (?, ?)`
Т.е. при обновлении удаляются вообще все привязки книги к жанрам, а потом вставляются и те, которые были и должны остаться, и новые, которых ещё не было. Мне кажется, что это нехорошо..

Поэтому пришлось городить огород с объединением существующего набора жанров и нового - самостоятельно из существующей коллекции удалять те, которых нет в новой, и добавлять те новые, которых в ней не было. На всякий случай перестраховался и сделал всё это дело на стримах с предикатами, сравнивающими по id, т.к. нет методов `equals` и `hashСode` у entities. Хотя, возможно, это уже и лишнее, т.к., насколько я понимаю, из persistence context возвращаются прямо в точности одинаковые объекты жанров, которые потенциально можно просто через `==` сопоставлять.
Получается в итоге вот так:
`Hibernate: update books set author_id=?, description=?, title=? where id=?`
`Hibernate: delete from books_genres where book_id=? and genre_id=?`
`Hibernate: delete from books_genres where book_id=? and genre_id=?`
`Hibernate: insert into books_genres (book_id, genre_id) values (?, ?)`

Может быть, желаемого результата при рекомендованном подходе можно добиться как-то проще?

MAMKUH6OPIIIUK committed:
Исправления по результатам первой итерации проверки ДЗ №7
1. Убрано ненужное управление версиями h2 в pom. Поднята версия
2. Аннотации `@Query` в репозитории книг заменены на `@EntityGraph` (с созданием соответствующего графа у сущности `Book`). Добавлено переопределение метода `findById` тоже с использованием графа
3. В `BookRequestDto` и `CommentRequestDto` добавлено поле id. Выполнены соответствующие изменения в сигнатурах методов сервисов, мапперах и тестах.
4. Изменена методика обновления книги - раньше конструировался и сохранялся новый объект книги, теперь изменяется существующий

MAMKUH6OPIIIUK committed:
Применил для комментариев ту же новую методику обновления, что и для книг (изменение существующего объекта).
Заодно убрал возможность перепривязки комментария к другой книге при обновлении, т.к. это выглядело странновато.

vlad-iv commented on spring-hw-07/src/main/java/ru/otus/spring/homework/oke/mapper/BookMapper.java:
```diff
@@ -0,0 +1,75 @@
+        book.setTitle(bookRequestDto.getTitle());
+        book.setDescription(bookRequestDto.getDescription());
+        book.setAuthor(newAuthor);
+        Set<Genre> bookGenres = book.getGenres();
```
Предлагаю вызвать setGenres(newGenres)

vlad-iv commented on spring-hw-07/src/main/java/ru/otus/spring/homework/oke/mapper/BookMapper.java:
```diff
@@ -0,0 +1,75 @@
+     * @return измененный в процессе работы аргумент book со всеми обновленными полями
+     */
+    public Book mergeBookInfo(Book book, BookRequestDto bookRequestDto, Author newAuthor,
+                              Collection<Genre> newGenres) {
```
Можно сразу Set передавать

vlad-iv commented on spring-hw-07/src/main/java/ru/otus/spring/homework/oke/mapper/BookMapper.java:
```diff
@@ -0,0 +1,75 @@
+
+    private final GenreMapper genreMapper;
+
+    public Book mapToBook(BookRequestDto bookRequestDto, Author author, Collection<Genre> genres) {
```
Можно сразу Set передавать

vlad-iv commented on spring-hw-07/src/main/java/ru/otus/spring/homework/oke/mapper/BookMapper.java:
```diff
@@ -0,0 +1,75 @@
+        book.setTitle(bookRequestDto.getTitle());
+        book.setDescription(bookRequestDto.getDescription());
+        book.setAuthor(author);
+        book.setGenres(genres.stream().collect(Collectors.toSet()));
```
Предлагаю вызвать setGenres(newGenres)

vlad-iv commented on spring-hw-07/src/main/java/ru/otus/spring/homework/oke/mapper/BookMapper.java:
```diff
@@ -0,0 +1,75 @@
+     * @param newGenres новая коллекция жанров книги (должны существовать)
+     * @return измененный в процессе работы аргумент book со всеми обновленными полями
+     */
+    public Book mergeBookInfo(Book book, BookRequestDto bookRequestDto, Author newAuthor,
```
Можно void

MAMKUH6OPIIIUK committed:
Исправления по результатам второй итерации проверки ДЗ №7
